js/src/components/google-account-card/useSwitchGoogleAccount.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { glaData } from '.~/constants';
+import { API_NAMESPACE } from '.~/data/constants';
+import useGoogleAuthorization from '.~/hooks/useGoogleAuthorization';
+import useDispatchCoreNotices from '.~/hooks/useDispatchCoreNotices';
+import useApiFetchCallback from '.~/hooks/useApiFetchCallback';
+
+/**
+ * A hook that returns a handler that initiates Google account disconnect and connect, to support switching to a different Google account.
+ *
+ * The `handleSwitch` handler is meant to be used in button click handler. Upon button click, the handler will:
+ *
+ * 1. Display an info notice that the process is running and request the users to wait.
+ * 2. Call `DELETE /google/connect` API to disconnect the existing connected Google account.
+ * 3. Call `GET /google/connect` API to get the Google OAuth URL.
+ * 4. Redirect the browser to the URL.
+ * 5. If there is an error in the above process, it will display an error notice.
+ *
+ * @return {Array} `[ handleSwitch, loading ]`
+ * 		- `handleSwitch` is meant to be used as button click handler.
+ * 		- `loading` is a state to indicate that the process is running.
+ */
+const useSwitchGoogleAccount = () => {
+	const pageName = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
+	const { createNotice, removeNotice } = useDispatchCoreNotices();
+	const [
+		fetchGoogleDisconnect,
+		{ loading: loadingGoogleDisconnect },
+	] = useApiFetchCallback( {
+		path: `${ API_NAMESPACE }/google/connect`,
+		method: 'DELETE',
+	} );
```

Comment:
❔ or 💅 

There's an action for disconnecting Google account. Are there any concerns? Or it could reuse the action here.

https://github.com/woocommerce/google-listings-and-ads/blob/7fb9c8799ab7f41789834f67c802ab396d03a305/js/src/data/actions.js#L413-L433

Reply:
Ahh yes, I tested that out. If we use that action, it would cause an update in the wp-data redux store, and the UI will change to the "Connect" card for a brief moment, before the browser redirects to the Google Auth page.

Reply:
I added code comment for this in [`02799c3` (#1072)](https://github.com/woocommerce/google-listings-and-ads/pull/1072/commits/02799c3b90957c7e98264562500a929556b0d0f9).

Reply:
💅 
We could add a parameter, e.g. `skipStoreInvalidation`, to the `disconnectGoogleAccount` action for skipping yielding the `TYPES.DISCONNECT_ACCOUNTS_GOOGLE`. Therefore, I think this hook would look more concise.

```js
const useSwitchGoogleAccount = () => {
	const pageName = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
	const { createNotice, removeNotice } = useDispatchCoreNotices();
	const { disconnectGoogleAccount } = useAppDispatch();
	const [ isRunning, setRunning ] = useState( false );
	const [ fetchGoogleConnect ] = useGoogleAuthorization( pageName );

	const handleSwitch = async () => {
		setRunning( true );

		const { notice } = await createNotice(
			'info',
			__(
				'Connecting to a different Google account, please wait…',
				'google-listings-and-ads'
			)
		);

		try {
			await disconnectGoogleAccount();
			const { url } = await fetchGoogleConnect( true );
			window.location.href = url;
		} catch ( error ) {
			setRunning( false );
			removeNotice( notice.id );
			createNotice(
				'error',
				__(
					'Unable to connect to a different Google account. Please try again later.',
					'google-listings-and-ads'
				)
			);
		}
	};

	return [ handleSwitch, isRunning ];
};
```

Reply:
> We could add a parameter, e.g. `skipStoreInvalidation`, to the `disconnectGoogleAccount` action for skipping yielding the `TYPES.DISCONNECT_ACCOUNTS_GOOGLE`.

I tried out the idea, and to be honest, personally I'm not really fond of that. The `skipStoreInvalidation` parameter is there to work with our wp-data technical implementation, and it does not really describe `disconnectGoogleAccount`. In other words, in layman term regardless of wp-data store, "I want to disconnect Google account, why is this skipStoreInvalidation needed?"

The other thing is that we would also need to handle it in the resolver part, like below:

```js
// add the new "! action.skipStoreInvalidation" condition below.
getGoogleAccount.shouldInvalidate = ( action ) => {
	return (
		action.type === TYPES.DISCONNECT_ACCOUNTS_GOOGLE &&
		! action.skipStoreInvalidation
	);
};
```

So to support this `skipStoreInvalidation`, we would need to modify `actions.js`, `resolvers.js` and the consuming hook here.

While it can make the hook look more concise, I think overall it also makes things harder to reason about and not as straightforward. So I'm leaning towards what we have now in here. 😄

Reply:
Thanks for replying. I got more of your thoughts. 🙏 

> The `skipStoreInvalidation` parameter is there to work with our wp-data technical implementation, and it does not really describe `disconnectGoogleAccount`. In other words, in layman term regardless of wp-data store, "I want to disconnect Google account, why is this skipStoreInvalidation needed?"

> I think overall it also makes things harder to reason about and not as straightforward.

I think it's understandable that wp-data is the main global state manager in this project, and that adding a control parameter based on its characteristics should be a reasonable approach. Because eventually, frontend maintainers need to understand how (internal implementation of) wp-data invalidates runtime cache state and drives it to update relevant state from the source. Otherwise, we might have more and more similar workarounds.

Even if `skipStoreInvalidation` may seem a bit odd, a comment explaining its purpose is enough to show why, just as a comment is needed in this hook to explain why. And as much as possible, the state management could be kept in `. ~/data` for better codebase consistency and predictability.

> So I'm leaning towards what we have now in here.

My main point (started with ❔ ) above was to get thoughts and discussions. I think it's ok to keep this now.